### PR TITLE
Tolerate a complete URL in the ID field

### PR DESCRIPTION
**Pull request by inductiveload**
This makes it a bit easier when copy-pasting from the address bar.

Bug: T271424

### CODE REVIEW

File: src/Controller/UploadController.php [hunk 1]
```diff
@@ -175,6 +175,21 @@ public function init( Request $request, Response $response ) {
 		], $response );
 	}
 
+	/**
+	 * Try to extract an IA ID from a string that might be a URL or ID
+	 * @param string $urlOrId IA item URL or ID
+	 * @return string the IA ID if we can figure one out
+	 */
+	private function extractIaId( $urlOrId ) {
+		if ( str_starts_with( $urlOrId, 'http' ) ) {
```
Comment (samwilson) [inline src/Controller/UploadController.php:184]: I think `str_starts_with()` is only PHP 8.

File: src/Controller/UploadController.php [hunk 2]
```diff
@@ -184,7 +199,7 @@ public function init( Request $request, Response $response ) {
 	public function fill( Request $request, Response $response ) {
 		// Get inputs.
 		$query = $request->getQueryParams();
-		$iaId = trim( $query['iaId'] ?? '' );
+		$iaId = $this->extractIaId( trim( $query['iaId'] ?? '' ) );
```
Comment (samwilson) [inline src/Controller/UploadController.php:202]: Maybe move `trim()` into the method itself.

File: src/Controller/UploadController.php [hunk 3]
```diff
@@ -175,6 +175,21 @@ public function init( Request $request, Response $response ) {
...
+	/**
+	 * Try to extract an IA ID from a string that might be a URL or ID
+	 * @param string $urlOrId IA item URL or ID
+	 * @return string the IA ID if we can figure one out
+	 */
+	private function extractIaId( $urlOrId ) {
+		if ( str_starts_with( $urlOrId, 'http' ) ) {
+			$path = parse_url( $urlOrId,  PHP_URL_PATH );
+			$pathParts = explode( '/', $path );
+			$urlOrId = $pathParts[2];
```
Comment (samwilson) [inline src/Controller/UploadController.php:187]: Probably worth checking if `$pathParts[2]` exists before using it.

File: src/Controller/UploadController.php [hunk 4]
```diff
@@ -175,6 +175,21 @@ public function init( Request $request, Response $response ) {
 		], $response );
 	}
 
+	/**
+	 * Try to extract an IA ID from a string that might be a URL or ID
+	 * @param string $urlOrId IA item URL or ID
+	 * @return string the IA ID if we can figure one out
+	 */
+	private function extractIaId( $urlOrId ) {
```
Comment (samwilson) [inline src/Controller/UploadController.php:183]: Add param and return type hints.

**Review by samwilson — COMMENTED** [review]
Great idea.

File: src/Controller/UploadController.php (hunk 2 above)
Comment (inductiveload) [reply src/Controller/UploadController.php]: OK

File: src/Controller/UploadController.php (hunk 3 above)
Comment (inductiveload) [reply src/Controller/UploadController.php]: Yeah.. I found that out the hard way

File: src/Controller/UploadController.php (hunk 4 above)
Comment (inductiveload) [reply src/Controller/UploadController.php]: OK

File: src/Controller/UploadController.php (hunk 1 above)
Comment (inductiveload) [reply src/Controller/UploadController.php]: Huh, true, but it works in `php:7.3-apache-stretch`. Anyhoo, `strpos` now.

**Pull request merged**